crates/hir-ty/src/infer/expr.rs

```diff
@@ -603,6 +603,11 @@ impl<'a> InferenceContext<'a> {
                             .build();
                         self.write_method_resolution(tgt_expr, func, subst.clone());
                     }
+                    let actual_ret = inner_ty.clone();
+                    let residual = self.resolve_ops_try_err();
+                    let expect_err = self.resolve_associated_type(self.return_ty.clone(), residual);
+                    let actual_err = self.resolve_associated_type(actual_ret, residual);
+                    self.table.unify(&actual_err, &expect_err);
```

Comment:
I think this would cause a false positive type mismatch in this code:
```Rust
fn x() -> Result<(), E2> {
    let a: Result<i32, _> = Ok(2);
    let b = a?;
    let c: Result<i32, E1> = a;
    Ok(())
}
```

Reply:
<img width="806" alt="截屏2023-02-17 20 43 46" src="https://user-images.githubusercontent.com/7943146/219656816-ef974c1e-e523-493d-8e45-3811575ecbcf.png">
<img width="651" alt="截屏2023-02-17 20 47 06" src="https://user-images.githubusercontent.com/7943146/219657210-238d1ef6-251c-4d25-8599-e69f1e396ac9.png">


I use rustc 1.66.1, it's a type mismatch and ra with the patch give a hint.

Reply:
With the from impl, it works ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=96c3b6e9877509886eab482f42254d20)).